### PR TITLE
Fixing JIRA: https://wso2.org/jira/browse/APIMANAGER-3963

### PR DESCRIPTION
**Pull request by chamindias**

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/statshandler/GatewayStatsUpdater.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+     * This method updates the stat publishing status in gateway domain
+     *
+     * @param receiverUrl      event receiver url
+     * @param user             username for the event receiver
+     * @param password         password for the event receiver
+     * @param statUpdateStatus status of the stat publishing state
+     */
+    public void updateStatPublishGateway(String receiverUrl, String user, String password, Boolean statUpdateStatus) {
+
+        log.debug("Updating stats publishing status in Gateway.");
```
Comment (nuwand) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/statshandler/GatewayStatsUpdater.java]: Need to check log.isDebugEnabled()

Comment (chamindias) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/statshandler/GatewayStatsUpdater.java]: PR for Checking whether debug option is enabled.
https://github.com/wso2/carbon-apimgt/pull/908

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/statshandler/GatewayStatsUpdater.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+        //updating stat publishing at the receiver node, self update
+        APIManagerAnalyticsConfiguration analyticsConfiguration = APIManagerAnalyticsConfiguration.getInstance();
+        analyticsConfiguration.setAnalyticsEnabled(statUpdateStatus);
+        log.debug("Updated stats publishing status in Gateway to : " + statUpdateStatus);
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/statshandler/GatewayStatsUpdater.java]: Check log.isDebugEnabled()

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/statshandler/GatewayStatsUpdater.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+                    //values related to stats publishing status are only updated if all of them are non-empty
+                    if (!(receiverUrl.isEmpty()) && !(user.isEmpty()) && !(password.isEmpty())) {
+                        analyticsConfiguration.setBamServerUrlGroups(receiverUrl);
+                        analyticsConfiguration.setBamServerUser(user);
+                        analyticsConfiguration.setBamServerPassword(password);
+
+                        eventingConfigData.setUrl(receiverUrl);
+                        eventingConfigData.setUserName(user);
+                        eventingConfigData.setPassword(password);
+                        log.debug("BAMServerURL : " + receiverUrl + " , BAMServerUserName : " + user + " , " +
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/statshandler/GatewayStatsUpdater.java]: Need to check log.isDebugEnabled()

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/statshandler/GatewayStatsUpdater.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        EventingConfigData eventingConfigData = null;
+        if (serviceDataPublisherAdmin != null) {
+            eventingConfigData = serviceDataPublisherAdmin.getEventingConfigData();
+        }
+
+        try {
+            if (eventingConfigData != null) {
+                //config values are updated if the stats publishing is true
+                if (statUpdateStatus) {
+                    log.debug("Updating values related to stats publishing status.");
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/statshandler/GatewayStatsUpdater.java]: Check log.isDebugEnabled()

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/statshandler/GatewayStatsUpdater.java (hunk 2 above)
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/statshandler/GatewayStatsUpdater.java]: Above mentioned pull request has fixed them. So please discard this.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/StatUpdateClusterMessage.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        return null;
+    }
+
+    @Override
+    public void execute(ConfigurationContext configurationContext) throws ClusteringFault {
+
+        //update the service variable, a boolean variable representing the stat data publishing in the node
+        APIManagerAnalyticsConfiguration instanceOfAPIAnalytics = APIManagerAnalyticsConfiguration.getInstance();
+        instanceOfAPIAnalytics.setAnalyticsEnabled(statUpdateStatus);
+        log.debug("Updated Stat publishing status to : " + statUpdateStatus);
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/StatUpdateClusterMessage.java]: Check log.isDebugEnabled()

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/StatUpdateStorePublisherDomain.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    private static final Log log = LogFactory.getLog(StatUpdateStorePublisherDomain.class);
+
+    /**
+     * This method updates the stat publishing status in Store-Publisher domain by using a cluster message
+     *
+     * @param statUpdateStatus status of the stat publishing state
+     */
+    public void updateStatsPublishStore(boolean statUpdateStatus) {
+
+        log.debug("Started updating Stats publishing status to : " + statUpdateStatus);
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/StatUpdateStorePublisherDomain.java]: Check log.isDebugEnabled()

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/StatUpdateStorePublisherDomain.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        if (clusteringAgent != null) {
+            //changing stat publishing status at other nodes via a cluster message
+            try {
+                clusteringAgent.sendMessage(new StatUpdateClusterMessage(statUpdateStatus), true);
+            } catch (ClusteringFault clusteringFault) {
+                //error is only logged because initially gateway has modified the status
+                log.error("Failed to send cluster message to Publisher/Store domain " +
+                        "and update stats publishing status.");
+            }
+            log.debug("Successfully updated Stats publishing status to : " + statUpdateStatus);
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/StatUpdateStorePublisherDomain.java]: Check log.isDebugEnabled()

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 8]
```diff
@@ -2419,6 +2424,54 @@ private boolean addExternalAPIStoresDetails(APIIdentifier apiId,Set<APIStore> ap
...
+                    password = "";
+                }
+                stub.updateStatPublishGateway(receiverUrl, user, password, updatedStatus);
+
+                //send cluster message to publisher-store domain
+                storePublisherMessageAgent.updateStatsPublishStore(updatedStatus);
+            }
+        } catch (AxisFault axisFault) {
+            //error is only logged because the process should be executed in all gateway environments
+            log.error("Error in calling Statsupdate web service in Gateway Environment." + axisFault);
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: Use log.error(msg, exception).

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 9]
```diff
@@ -2419,6 +2424,54 @@ private boolean addExternalAPIStoresDetails(APIIdentifier apiId,Set<APIStore> ap
...
+        } catch (RemoteException remoteException) {
+            //error is only logged because the change is affected in gateway environments,
+            // and the process should be executed in all environments
+            log.error("Error in updating Stats publish status in Gataways. " + remoteException);
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: Use log.error(msg, exception).

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/StatUpdateStorePublisherDomain.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+        ClusteringAgent clusteringAgent = ServiceReferenceHolder.getContextService().getServerConfigContext().
+                                            getAxisConfiguration().getClusteringAgent();
+
+        if (clusteringAgent != null) {
+            //changing stat publishing status at other nodes via a cluster message
+            try {
+                clusteringAgent.sendMessage(new StatUpdateClusterMessage(statUpdateStatus), true);
+            } catch (ClusteringFault clusteringFault) {
+                //error is only logged because initially gateway has modified the status
+                log.error("Failed to send cluster message to Publisher/Store domain " +
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/StatUpdateStorePublisherDomain.java]: Use log.error(msg, exception) and logging the exception is missing

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/statshandler/GatewayStatsUpdater.java (hunk 2 above)
Comment (chamindias) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/statshandler/GatewayStatsUpdater.java]: Fixed in : https://github.com/wso2/carbon-apimgt/pull/908/

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/StatUpdateClusterMessage.java (hunk 5 above)
Comment (chamindias) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/StatUpdateClusterMessage.java]: Fixed in : https://github.com/wso2/carbon-apimgt/pull/908/

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/StatUpdateStorePublisherDomain.java (hunk 6 above)
Comment (chamindias) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/StatUpdateStorePublisherDomain.java]: Please note this "StatUpdateStorePublisherDomain" class has been removed and the methods are included in "APIProviderImpl". In APIProviderImpl, log.isDebugEnabled() is checked.
Please refer : https://github.com/wso2/carbon-apimgt/commit/4df13636996d6ffab9a90a0a5ff9dca53c2e04e8

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/StatUpdateStorePublisherDomain.java (hunk 7 above)
Comment (chamindias) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/StatUpdateStorePublisherDomain.java]: Please note this "StatUpdateStorePublisherDomain" class has been removed and the methods are included in "APIProviderImpl". In APIProviderImpl, log.isDebugEnabled() is checked.
Please refer : https://github.com/wso2/carbon-apimgt/commit/4df13636996d6ffab9a90a0a5ff9dca53c2e04e8 (removing class and relocating methods) and https://github.com/wso2/carbon-apimgt/commit/14ad7f3bee29148abf11916998e284fa70d59fcb (check debug option)

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java (hunk 8 above)
Comment (chamindias) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: This is fixed in : https://github.com/wso2/carbon-apimgt/commit/2917bb535a4686418310b787af361524ff352252 (APIProviderImpl class)

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java (hunk 9 above)
Comment (chamindias) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: This is fixed in : https://github.com/wso2/carbon-apimgt/commit/2917bb535a4686418310b787af361524ff352252 (APIProviderImpl class)

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/StatUpdateStorePublisherDomain.java (hunk 10 above)
Comment (chamindias) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/StatUpdateStorePublisherDomain.java]: Please note this "StatUpdateStorePublisherDomain" class has been removed (https://github.com/wso2/carbon-apimgt/commit/4df13636996d6ffab9a90a0a5ff9dca53c2e04e8) and the methods are included in "APIProviderImpl". In APIProviderImpl, this is fixed in : https://github.com/wso2/carbon-apimgt/commit/14ad7f3bee29148abf11916998e284fa70d59fcb

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 11]
```diff
@@ -2419,6 +2424,54 @@ private boolean addExternalAPIStoresDetails(APIIdentifier apiId,Set<APIStore> ap
...
+
+            for (String gatewayEnvironmentName : gatewyEnvironments.keySet()) {
+                Environment currentGatewayEnvironment = gatewyEnvironments.get(gatewayEnvironmentName);
+                String gatewayServiceUrl = currentGatewayEnvironment.getServerURL();
+                String gatewayUserName = currentGatewayEnvironment.getUserName();
+                String gatewayPassword = currentGatewayEnvironment.getPassword();
+
+                //get the stub and the call the admin service with the credentials
+                GatewayStatsUpdateServiceStub stub =
+                        new GatewayStatsUpdateServiceStub(gatewayServiceUrl + "GatewayStatsUpdateService");
```
Comment (nuwand) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: Use a constant for "GatewayStatsUpdateService"

Comment (chamindias) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: Done in : https://github.com/chamindias/carbon-apimgt/commit/2917bb535a4686418310b787af361524ff352252

**Pull request merged**